### PR TITLE
Enforce type-safety when comparing query result values

### PR DESCRIPTION
**Pull request by sergey-melnychuk**
Throw an `IllegalArgumentException` when comparing apples to oranges.
IMHO this must significantly reduce [WTFPM](http://3.bp.blogspot.com/-ilMjE1Gh3Yg/VpUAmd-6TWI/AAAAAAAAAbg/-FJ08zxN42s/s1600/WFTPM.png) rate when troubleshooting assertions in Trino product tests.

### CODE REVIEW

File: tempto-core/src/main/java/io/trino/tempto/assertions/QueryResultValueComparator.java [hunk 1]
```diff
@@ -262,7 +260,7 @@ private boolean timestampWithTimezoneEqual(Object actual, Object expected)
         if (actual instanceof ZonedDateTime && expected instanceof ZonedDateTime) {
             return actual.equals(expected);
         }
-        return false;
+        return fail(actual, expected);
```
Comment (findepi) [inline tempto-core/src/main/java/io/trino/tempto/assertions/QueryResultValueComparator.java]: `return fail` -> `throw incompatibleTypes(...)`

File: tempto-core/src/main/java/io/trino/tempto/assertions/QueryResultValueComparator.java [hunk 2]
```diff
@@ -284,4 +282,11 @@ private static double getDoubleValue(Object object)
     {
         return ((Number) object).doubleValue();
     }
+
+    // Make compiler happy: declare boolean is returned to avoid dead code 'return false;' after each 'fail' call.
+    private static boolean fail(Object actual, Object expected) {
+        String message = format("Invalid comparison: 'actual' value is of type '%s' while 'expected' value is of type '%s'",
+                actual.getClass().getCanonicalName(), expected.getClass().getCanonicalName());
```
Comment (findepi) [inline tempto-core/src/main/java/io/trino/tempto/assertions/QueryResultValueComparator.java]: Can actual or expected be null at this point?

File: tempto-core/src/main/java/io/trino/tempto/assertions/QueryResultValueComparator.java (hunk 1 above)
Comment (sergey-melnychuk) [reply tempto-core/src/main/java/io/trino/tempto/assertions/QueryResultValueComparator.java]: done

File: tempto-core/src/main/java/io/trino/tempto/assertions/QueryResultValueComparator.java (hunk 2 above)
Comment (sergey-melnychuk) [reply tempto-core/src/main/java/io/trino/tempto/assertions/QueryResultValueComparator.java]: made it null-safe

File: tempto-core/src/main/java/io/trino/tempto/assertions/QueryResultValueComparator.java [hunk 3]
```diff
@@ -284,4 +282,11 @@ private static double getDoubleValue(Object object)
     {
         return ((Number) object).doubleValue();
     }
+
+    private static IllegalArgumentException incompatibleTypes(Object actual, Object expected) {
+        String message = format("Type mismatch: 'actual' value is of type '%s' while 'expected' value is of type '%s'",
+                requireNonNull(actual, "'actual' value is null").getClass().getCanonicalName(),
+                requireNonNull(expected, "'expected' value is null").getClass().getCanonicalName());
```
Comment (findepi) [inline tempto-core/src/main/java/io/trino/tempto/assertions/QueryResultValueComparator.java:289]: why not `getName()`?

**Pull request merged**